File: frontend/src/types/scan.ts
/**
 * 扫描批次相关 TypeScript 类型定义
 */

export interface ScanBatch {
  id: number
  batch_name: string | null
  uploaded_by: number | null
  uploaded_at: string
  scan_started_at: string | null
  scan_finished_at: string | null
  file_size_bytes: number | null
  total_hosts: number | null
  new_count: number
  changed_count: number
  missing_count: number
  status: 'pending' | 'confirmed' | 'rejected'
}

export interface ScanBatchListResponse {
  items: ScanBatch[]
  total: number
}

export interface ScanConfirmRequest {
  new_assets: Array<{
    ip_address: string
    asset_type: string
    location: string
    owner: string
    business_system: string
    importance: string
    network_zone: string
    asset_no?: string
    mac_address?: string
    hostname?: string
    os_info?: string
  }>
}

// ── 差异详情类型 ──────────────────────────────────────────

export interface DiffPort {
  port_number: number
  protocol: string
  service_name: string | null
  service_version: string | null
  state: string | null
}

export interface DiffPortChange {
  port_number: number
  protocol: string
  old_service: string | null
  new_service: string | null
  old_version: string | null
  new_version: string | null
  old_state: string | null
  new_state: string | null
  change_type: 'added' | 'removed' | 'modified'
}

export interface DiffNewHost {
  ip_address: string
  mac_address: string | null
  hostname: string | null
  os_info: string | null
  ports: DiffPort[]
}

export interface DiffChangedHost {
  ip_address: string
  mac_address: string | null
  hostname: string | null
  os_info: string | null
  matched_asset_id: number | null
  matched_asset_no: string | null
  port_changes: DiffPortChange[]
  current_ports: DiffPort[]
  scan_ports: DiffPort[]
}

export interface DiffMissingHost {
  ip_address: string
  mac_address: string | null
  hostname: string | null
  matched_asset_id: number | null
  matched_asset_no: string | null
  missing_count: number
}

export interface ScanDiffResponse {
  batch_id: number
  batch_name: string | null
  status: string
  total_hosts: number
  new_count: number
  changed_count: number
  missing_count: number
  new_hosts: DiffNewHost[]
  changed_hosts: DiffChangedHost[]
  missing_hosts: DiffMissingHost[]
}

File: backend/app/schemas/scan.py
"""
Scan batch related Pydantic v2 schemas
"""
from datetime import datetime

from pydantic import BaseModel


class ScanBatchRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    batch_name: str | None
    uploaded_by: int | None
    uploaded_at: datetime
    scan_started_at: datetime | None
    scan_finished_at: datetime | None
    file_size_bytes: int | None
    total_hosts: int | None
    new_count: int
    changed_count: int
    missing_count: int
    status: str


class ScanBatchListResponse(BaseModel):
    items: list[ScanBatchRead]
    total: int


class ScanConfirmRequest(BaseModel):
    """Confirm import request: contains supplementary fields for new assets"""
    new_assets: list[dict] = []


# ── Diff detail Schemas ─────────────────────────────────────────


class DiffPortRead(BaseModel):
    """Port-level diff"""
    port_number: int
    protocol: str
    service_name: str | None = None
    service_version: str | None = None
    state: str | None = None


class DiffHostPortChange(BaseModel):
    """Port change detail"""
    port_number: int
    protocol: str
    old_service: str | None = None
    new_service: str | None = None
    old_version: str | None = None
    new_version: str | None = None
    old_state: str | None = None
    new_state: str | None = None
    change_type: str = ""  # added / removed / modified


class DiffNewHost(BaseModel):
    """Newly discovered host"""
    ip_address: str
    mac_address: str | None = None
    hostname: str | None = None
    os_info: str | None = None
    ports: list[DiffPortRead] = []


class DiffChangedHost(BaseModel):
    """Changed host"""
    ip_address: str
    mac_address: str | None = None
    hostname: str | None = None
    os_info: str | None = None
    matched_asset_id: int | None = None
    matched_asset_no: str | None = None
    port_changes: list[DiffHostPortChange] = []
    current_ports: list[DiffPortRead] = []
    scan_ports: list[DiffPortRead] = []


class DiffMissingHost(BaseModel):
    """Missing host"""
    ip_address: str
    mac_address: str | None = None
    hostname: str | None = None
    matched_asset_id: int | None = None
    matched_asset_no: str | None = None
    missing_count: int = 0


class ScanDiffResponse(BaseModel):
    """Scan batch diff detail"""
    batch_id: int
    batch_name: str | None = None
    status: str
    total_hosts: int
    new_count: int
    changed_count: int
    missing_count: int
    new_hosts: list[DiffNewHost] = []
    changed_hosts: list[DiffChangedHost] = []
    missing_hosts: list[DiffMissingHost] = []
